Alert Subsys to OOP and SNMPTraps trigger Alert Rules

h-barnhart opened the pull request:
DO NOT DELETE THIS TEXT

Changed alerting subsystem to OOP (see file changes). SNMP traps and ICMP device down events will trigger alerts immediately rather than wait for next polling cycle.

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [Yes] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 9765`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

murrant commented:
RunRules should be run for every "defined" trap.  So you don't have to specify it in everyone.  Probably here: LibreNMS/Snmptrap/Dispatcher.php

h-barnhart commented:
I thought about that, but went the other way incase there was a trap that
wouldn't run rules. I can put it back though. Still trying to figure out
how to get it to pass phpunit tests.

On Thu, Feb 7, 2019 at 11:31 AM Tony Murray <notifications@github.com>
wrote:

> RunRules should be run for every "defined" trap. So you don't have to
> specify it in everyone. Probably here: LibreNMS/Snmptrap/Dispatcher.php
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/librenms/librenms/pull/9765#issuecomment-461522768>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AqjJ32jgJA88RdoHXOAS6tXuiJ33zZHRks5vLGMKgaJpZM4acrPT>
> .
>


-- 
Sincerely,
Heath Barnhart
Network Administrator
[image: KanREN] <http://www.kanren.net/>
[image: phone] 785-856-9815
2029 Becker Drive, Suite 282
Lawrence, Kansas 66047
[image: linkedin]
<https://www.linkedin.com/company/92399?trk=tyah&trkInfo=clickedVertical%3Acompany%2CclickedEntityId%3A92399%2Cidx%3A1-1-1%2CtarId%3A1440002635645%2Ctas%3AKanREN>
 [image: twitter] <https://twitter.com/TheKanREN> [image: twitter]
<http://www.kanren.net/feed/> need support? <support@kanren.net>

murrant commented:
People may even want to alert on just basic alerts adding logging to the eventlog.

murrant commented:
Tests don't use init.php (for whatever reason)

h-barnhart commented:
Moved RunRules() to Dispatcher.php so it will run rules for any handled SNMP trap. I added it as an else to the fallback logic, so if no handler is found RunRules() isn't called. As far as I can tell it works, but doesn't pass unit tests because the test env doesn't have access to the database.

Included the function files functions.php and alerts.inc.php to SnmpTrapTests.php so it will have access to RunRules(). If pre-commit.php is run with the --db option it will check the database.

h-barnhart commented:
Not sure what to do at this point. Running pre-commit -u --db on my local server with the changes I've made passes.

murrant commented:
What I usually do is port the code to object oriented style and redirect the original function.  Examples: display() and is_valid_hostname()

h-barnhart commented:
Ok, I think understand what you are doing. I'll give it a shot. Unless you say otherwise I'm going to drop it in LibreNMS/Alerts.

Side note, not sure if its GitHub or Gmail, but most of your last comment got cut off in email update to the PR.

murrant commented:
@h-barnhart It's me. I type stuff to quickly then realize I forgot something, then edit the comment.  I should probably just create additional comments in those situations...

murrant commented:
I have updated the trap testing, so maybe we can re-visit this.

h-barnhart commented:
I've been busy at my job for the last month and haven't had much time to
work on this. When I get a chance I'll update my dev box and see what
happens.

On Wed, Mar 13, 2019 at 12:05 AM Tony Murray <notifications@github.com>
wrote:

> I have updated the trap testing, so maybe we can re-visit this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/librenms/librenms/pull/9765#issuecomment-472281314>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AqjJ36Jt4cq50dh-M0T4jiNxQ_Tu6mn0ks5vWIclgaJpZM4acrPT>
> .
>


-- 
Sincerely,
Heath Barnhart
Network Administrator
[image: KanREN] <http://www.kanren.net/>
[image: phone] 785-856-9815
2029 Becker Drive, Suite 282
Lawrence, Kansas 66047
[image: linkedin]
<https://www.linkedin.com/company/92399?trk=tyah&trkInfo=clickedVertical%3Acompany%2CclickedEntityId%3A92399%2Cidx%3A1-1-1%2CtarId%3A1440002635645%2Ctas%3AKanREN>
 [image: twitter] <https://twitter.com/TheKanREN> [image: twitter]
<http://www.kanren.net/feed/> need support? <support@kanren.net>

h-barnhart commented:
I saw the change to how logging is done, did that allow SNMP Trap logs to trigger alerts?

laf commented on LibreNMS/Alert/AlertRules.php:
```diff
@@ -0,0 +1,111 @@
+    {
+
+        //Check to see if under maintenance
+        if (IsMaintenance($device_id) > 0) {
```
This is failing tests: https://travis-ci.com/librenms/librenms/jobs/189340179

`$this->IsMaintenance()`?

laf replied:
No worries, I've edited the title to WIP so we know. Remove it when you're ready

h-barnhart commented:
A _little_ scope creep, in addition to getting traps to trigger ruleRun() I converted the rest of the alert subsytem to object oriented.

h-barnhart commented:
This is ready for review.

laf commented:
I'm running the code with a test rule to see if anything breaks.

laf commented:
```
PHP Fatal error:  require_once(): Failed opening required '/opt/librenms/includes/alerts.inc.php' (include_path='.:/usr/share/pear:/usr/share/php') in /opt/librenms/includes/init.php on line 90
PHP Stack trace:
PHP   1. {main}() /opt/librenms/alerts.php:0
PHP   2. require() /opt/librenms/alerts.php:34

Fatal error: require_once(): Failed opening required '/opt/librenms/includes/alerts.inc.php' (include_path='.:/usr/share/pear:/usr/share/php') in /opt/librenms/includes/init.php on line 90

Call Stack:
    0.0004     394512   1. {main}() /opt/librenms/alerts.php:0
    0.0008     418608   2. require('/opt/librenms/includes/init.php') /opt/librenms/alerts.php:34

```

h-barnhart commented:
Wonder if something got changed in a rebase, I'll check on it in the
morning. Thanks for testing it.

On Thu, May 2, 2019, 4:51 PM Neil Lathwood <notifications@github.com> wrote:

> PHP Fatal error:  require_once(): Failed opening required '/opt/librenms/includes/alerts.inc.php' (include_path='.:/usr/share/pear:/usr/share/php') in /opt/librenms/includes/init.php on line 90
> PHP Stack trace:
> PHP   1. {main}() /opt/librenms/alerts.php:0
> PHP   2. require() /opt/librenms/alerts.php:34
>
> Fatal error: require_once(): Failed opening required '/opt/librenms/includes/alerts.inc.php' (include_path='.:/usr/share/pear:/usr/share/php') in /opt/librenms/includes/init.php on line 90
>
> Call Stack:
>     0.0004     394512   1. {main}() /opt/librenms/alerts.php:0
>     0.0008     418608   2. require('/opt/librenms/includes/init.php') /opt/librenms/alerts.php:34
>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/librenms/librenms/pull/9765#issuecomment-488845517>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKUMTX5GWXJ6GT4QY442V3TPTNO6ZANCNFSM4GTSWPJQ>
> .
>

h-barnhart commented:
Looks like I forgot to include changes to init.php. I'll have in fixed shortly.

h-barnhart commented:
Should work now.

h-barnhart commented:
Spoke too soon, looks like something else got changed.

h-barnhart commented:
Ok, should be good to test now.

laf commented:
Re-testing now.

murrant commented on LibreNMS/Alert/AlertUtil.php:
```diff
@@ -41,4 +70,158 @@ public static function getDefaultAlertTransports()
+        }
+        if (strstr($rule, "%macros.")) {
+            if (++$x < 30) {
+                $rule = runMacros($rule, $x);
```
Undefined function.

h-barnhart replied:
Fixed 0a6260b

murrant commented on LibreNMS/Alert/AlertDB.php:
```diff
@@ -0,0 +1,118 @@
+        if ($query_builder) {
+            return QueryBuilderParser::fromJson($query_builder)->toSql();
+        } else {
+            return genSQLOld($rule);
```
Undefined function.

h-barnhart replied:
Fixed in 0a6260b.

h-barnhart commented:
Pushing up the fixes for the undefined functions. Not sure if its just my environment, but after switching to the new poller service icmp device down detected by ICMP isn't throwing an alert. I see the eventlog message so I know the ICMP check occurred, however, the alert is not thrown until the next polling cycle.

h-barnhart commented:
Fixed ICMP not triggering alert. Had it before the device status was written to the DB. Assuming code checks out should be ready for review and testing.

murrant commented on LibreNMS/Alert/RunAlerts.php:
```diff
@@ -0,0 +1,591 @@
+     * Run all alerts
+     * @return void
+     */
+    public function runAlerts()
```
This should be __construct()

h-barnhart replied:
Fixed

murrant commented:
Looking pretty close.  I'll give this a test.  Perhaps we should make a call for other testers in Discord.

murrant commented:
You'll need to incorporate #10253

h-barnhart commented:
Any further requests for changes?

murrant commented:
Just needs testing really.

laf commented:
Running tests again on this now.
